python/ray/tune/utils/resource_updater.py

```diff
@@ -0,0 +1,133 @@
+            # NOTE: This hides the possibility that Ray may be waiting for
+            # clients to connect.
+            resources.setdefault("CPU", 0)
+            resources.setdefault("GPU", 0)
+            logger.warning(
+                "Cluster resources cannot be detected or are 0. "
+                "You can resume this experiment by passing in "
+                "`resume=True` to `run`."
+            )
+
+        resources = resources.copy()
+        num_cpus = resources.pop("CPU", 0)
+        num_gpus = resources.pop("GPU", 0)
+        memory = ray_constants.from_memory_units(resources.pop("memory", 0))
+        object_store_memory = ray_constants.from_memory_units(
+            resources.pop("object_store_memory", 0)
+        )
+        custom_resources = resources
+
+        self._avail_resources = Resources(
+            int(num_cpus),
+            int(num_gpus),
+            memory=int(memory),
+            object_store_memory=int(object_store_memory),
+            custom_resources=custom_resources,
+        )
+        self._last_resource_refresh = time.time()
+        self._resources_initialized = True
+
+    def debug_string(self, total_resources: Dict[str, Any]) -> str:
+        """Returns a human readable message for printing to the console."""
+        if self._resources_initialized:
+            status = (
+                "Resources requested: {}/{} CPUs, {}/{} GPUs, "
+                "{}/{} GiB heap, {}/{} GiB objects".format(
+                    total_resources.pop("CPU", 0),
+                    self._avail_resources.cpu,
+                    total_resources.pop("GPU", 0),
+                    self._avail_resources.gpu,
+                    _to_gb(total_resources.pop("memory", 0.0)),
+                    _to_gb(self._avail_resources.memory),
+                    _to_gb(total_resources.pop("object_store_memory", 0.0)),
+                    _to_gb(self._avail_resources.object_store_memory),
+                )
+            )
+            customs = ", ".join(
+                [
+                    "{}/{} {}".format(
+                        total_resources.get(name, 0.0),
+                        self._avail_resources.get_res_total(name),
+                        name,
+                    )
+                    for name in self._avail_resources.custom_resources
+                    if not name.startswith(NODE_ID_PREFIX)
+                    and (total_resources.get(name, 0.0) > 0 or "_group_" not in name)
+                ]
+            )
+            if customs:
+                status += " ({})".format(customs)
+            return status
+        else:
+            return "Resources requested: ?"
+
+    def get_num_cpus(self) -> int:
+        self.update_avail_resources()
+        return self._avail_resources.cpu
+
+    def get_num_gpus(self) -> int:
+        self.update_avail_resources()
+        return self._avail_resources.cpu
+
+    def has_gpus(self) -> bool:
+        if self._resources_initialized:
```

Comment:
I wonder why the original code is gated by this.

Reply:
yes, it seems buggy to me.

Reply:
@richardliaw do you know why we did this for `has_gpus` in #5948?

Reply:
after discussion, the implementation of the current PR is correct.